plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java

```diff
@@ -94,7 +116,51 @@ private ConnectorSplit toIcebergSplit(FileScanTask task)
                 task.length(),
                 task.file().fileSizeInBytes(),
                 task.file().format(),
-                ImmutableList.of(),
+                getAddresses(task),
                 getPartitionKeys(task));
     }
+
+    private List<HostAddress> getAddresses(FileScanTask task)
+    {
+        if (!fetchSplitLocations) {
+            return ImmutableList.of();
+        }
+        return blockLocations(task);
+    }
+
+    private List<HostAddress> blockLocations(FileScanTask task)
+    {
+        Set<HostAddress> locations = new HashSet<>();
+        Path path = new Path(task.file().path().toString());
+        try {
+            FileSystem fs = hdfsEnvironment.getFileSystem(hdfsContext, path);
+            for (BlockLocation blockLocation : fs.getFileBlockLocations(path, task.start(), task.length())) {
+                locations.addAll(getHostAddresses(blockLocation));
+            }
+        }
+        catch (IOException e) {
+            log.warn("Failed to get block locations for path %s", path, e);
+        }
+        return new ArrayList<>(locations);
+    }
+
+    private static List<HostAddress> getHostAddresses(BlockLocation blockLocation)
+    {
+        // Hadoop FileSystem returns "localhost" as a default
+        return Arrays.stream(getBlockHosts(blockLocation))
+                .map(HostAddress::fromString)
+                .filter(address -> !"localhost".equals(address.getHostText()))
+                .collect(toImmutableList());
+    }
+
+    private static String[] getBlockHosts(BlockLocation blockLocation)
+    {
+        // To handle IOException from BlockLocation#getHosts
+        try {
+            return blockLocation.getHosts();
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
```

Comment:
this isn't converted to `TrinoException` up the call tree (unlike Hive connector). So, I'd recommend inlining this method and catch along with other IOExceptions in `blockLocations`.